src/java.base/share/classes/java/lang/Double.java

```diff
@@ -707,51 +709,76 @@ public static String toHexString(double d) {
             // For infinity and NaN, use the decimal output.
             return Double.toString(d);
         else {
-            // Initialized to maximum size of output.
-            StringBuilder answer = new StringBuilder(24);
-
-            if (Math.copySign(1.0, d) == -1.0)    // value is negative,
-                answer.append("-");                  // so append sign info
-
-            answer.append("0x");
-
+            boolean negative = Math.copySign(1.0, d) == -1.0;
             d = Math.abs(d);
 
-            if(d == 0.0) {
-                answer.append("0.0p0");
+            if (d == 0.0) {
+                return negative ? "-0x0.0p0" : "0x0.0p0";
             } else {
+                // Check if the value is subnormal (less than the smallest normal value)
                 boolean subnormal = (d < Double.MIN_NORMAL);
 
                 // Isolate significand bits and OR in a high-order bit
-                // so that the string representation has a known
-                // length.
+                // so that the string representation has a known length.
+                // This ensures we always have 13 hex digits to work with (52 bits / 4 bits per hex digit)
                 long signifBits = (Double.doubleToLongBits(d)
-                                   & DoubleConsts.SIGNIF_BIT_MASK) |
-                    0x1000000000000000L;
+                        & DoubleConsts.SIGNIF_BIT_MASK) |
+                        0x1000000000000000L;
+
+                // Calculate the number of trailing zeros in the significand (in groups of 4 bits)
+                // This is used to remove trailing zeros from the hex representation
+                // We limit to 12 because we want to keep at least 1 hex digit (13 total - 12 = 1)
+                int trailingZeros = Math.min(12, ((Long.numberOfTrailingZeros(signifBits) & 0xFC) >> 2));
```

Comment:
Some additional discussion of the computation here would be helpful to casual readers.

I would also be tempted to add

assert 0 <= trailingZeros && trailingZeros <= 12;